### PR TITLE
[Proposal] Add support for WithCancellation to the AsyncPageable

### PR DESCRIPTION
**Pull request by AlexanderSher**

### CODE REVIEW

File: sdk/core/Azure.Core/tests/PageableHelpersTests.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+using System.Threading.Tasks;
+using Azure.Core.TestFramework;
+using NUnit.Framework;
+
+namespace Azure.Core.Tests
+{
+    public class PageableHelpersTests
+    {
+        [Test]
+        public void PageableHelpers_CreateAsyncPageableFromDelegates()
```
Comment (AlexanderSher) [inline sdk/core/Azure.Core/tests/PageableHelpersTests.cs:17]: In this test, is created from two delegates. It fails to throw `OperationCanceledException` for the token that is provided by `WithCancellation`

File: sdk/core/Azure.Core/tests/PageableHelpersTests.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            var enumerable = CreateAsyncPageableFromDelegates(cts).WithCancellation(cts.Token);
+
+            Assert.CatchAsync<OperationCanceledException>(async () =>
+            {
+                await foreach (var i in enumerable) { }
+            });
+        }
+
+        [Test]
+        public void PageableHelpers_CreateAsyncPageableFromEnumerable()
```
Comment (AlexanderSher) [inline sdk/core/Azure.Core/tests/PageableHelpersTests.cs:29]: In this test, `AsyncPageable` is created from `IAsyncEnumerable` which passes `CancellationToken` into enumerator.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs [hunk 3]
```diff
@@ -662,40 +575,25 @@ async Task<Page<IncomingRelationship>> NextPageFunc(string nextLink, int? pageSi
         /// See the asynchronous version of this method for examples.
         /// </seealso>
         public virtual Pageable<IncomingRelationship> GetIncomingRelationships(string digitalTwinId, CancellationToken cancellationToken = default)
+            => PageableHelpers.CreateEnumerable(GetIncomingRelationshipsImplAsync(false, digitalTwinId, cancellationToken).EnsureSyncEnumerable(),
+                _clientDiagnostics, "DigitalTwinsGeneratedClient.ListIncomingRelationships");
+
+        private async IAsyncEnumerable<Page<IncomingRelationship>> GetIncomingRelationshipsImplAsync(bool async, string digitalTwinId, [EnumeratorCancellation] CancellationToken cancellationToken)
```
Comment (AlexanderSher) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:581]: Creating `Pageable` and `AsyncPageable` can be created from the same method that handles both `sync` and `async` code paths, which reduces the amount of copy/paste.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs [hunk 4]
```diff
@@ -662,40 +575,25 @@ async Task<Page<IncomingRelationship>> NextPageFunc(string nextLink, int? pageSi
...
-                }
-                catch (Exception e)
-                {
-                    scope.Failed(e);
-                    throw;
-                }
-            }
+            Response<IncomingRelationshipCollection> response = async
+                ? await _dtRestClient.ListIncomingRelationshipsAsync(digitalTwinId, cancellationToken).ConfigureAwait(false)
+                : _dtRestClient.ListIncomingRelationships(digitalTwinId, cancellationToken);
```
Comment (AlexanderSher) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:585]: This conditional logic can be codegen'ed which would simplify handwritten code even further.

File: sdk/core/Azure.Core/src/Shared/AutoRest/PageableHelpers.cs [hunk 5]
```diff
@@ -4,13 +4,26 @@
 #nullable enable
 
 using System;
+using System.Collections;
 using System.Collections.Generic;
+using System.Threading;
 using System.Threading.Tasks;
+using Azure.Core.Pipeline;
 
 namespace Azure.Core
```
Comment (pakrym) [inline sdk/core/Azure.Core/src/Shared/AutoRest/PageableHelpers.cs:13]: This type is to support code generation and the original copy is in https://github.com/Azure/autorest.csharp/blob/feature/v3/src/assets/Generator.Shared/PageableHelpers.cs 

Automatic CI pipeline overwrites this file when generator is updated so changes to it should be done in autorest.csharp repo.

**Comment by pakrym** [comment]
Might be a good time to do https://github.com/Azure/azure-sdk-for-net/issues/9896 and get rid of https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Shared/PageResponseEnumerator.cs

File: sdk/core/Azure.Core/src/Shared/AutoRest/PageableHelpers.cs (hunk 5 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/AutoRest/PageableHelpers.cs:13]: Ok, thanks.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs (hunk 3 above)
Comment (pakrym) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:581]: I'm not sure about merging the method into one. Seems like most size reduction is coming from removing scoping logic.

File: sdk/core/Azure.Core/src/Shared/AutoRest/PageableHelpers.cs (hunk 5 above)
Comment (pakrym) [reply sdk/core/Azure.Core/src/Shared/AutoRest/PageableHelpers.cs:13]: Let's discuss it here and move to autorest after. It's easier to see the impact in this repo.

**Comment by pakrym** [comment]
I like everything except merging sync and async enumerators into one. 

``` C#
   private async IAsyncEnumerable<Page<string>> GetRelationshipsImplAsync(bool async, string digitalTwinId, string relationshipName, [EnumeratorCancellation] CancellationToken cancellationToken)
        {
            Argument.AssertNotNull(digitalTwinId, nameof(digitalTwinId));

            Response<RelationshipCollection> response = async
                ? await _dtRestClient.ListRelationshipsAsync(digitalTwinId, relationshipName, cancellationToken).ConfigureAwait(false)
                : _dtRestClient.ListRelationships(digitalTwinId, relationshipName, cancellationToken);

            yield return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());

            while (!string.IsNullOrEmpty(response.Value.NextLink))
            {
                response = async
                    ? await _dtRestClient.ListRelationshipsNextPageAsync(response.Value.NextLink, digitalTwinId, relationshipName, cancellationToken).ConfigureAwait(false)
                    : _dtRestClient.ListRelationshipsNextPage(response.Value.NextLink, digitalTwinId, relationshipName, cancellationToken);

                yield return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());
            }
        }
```

``` C#
   private IEnumerable<Page<string>> GetRelationshipsImpl(bool async, string digitalTwinId, string relationshipName)
        {
            Argument.AssertNotNull(digitalTwinId, nameof(digitalTwinId));

            RelationshipCollection response = _dtRestClient.ListRelationships(digitalTwinId, relationshipName, cancellationToken);

            yield return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());

            while (!string.IsNullOrEmpty(response.Value.NextLink))
            {
                response = _dtRestClient.ListRelationshipsNextPage(response.Value.NextLink, digitalTwinId, relationshipName, cancellationToken);

                yield return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());
            }
        }

    private async IEnumerable<Page<string>> GetRelationshipsImplAsync(bool async, string digitalTwinId, string relationshipName)
        {
            Argument.AssertNotNull(digitalTwinId, nameof(digitalTwinId));

            RelationshipCollection response = _dtRestClient.ListRelationshipsAsync(digitalTwinId, relationshipName, cancellationToken);

            yield return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());

            while (!string.IsNullOrEmpty(response.Value.NextLink))
            {
                response = await _dtRestClient.ListRelationshipsNextPageAsync(response.Value.NextLink, digitalTwinId, relationshipName, cancellationToken);

                yield return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());
            }
        }
```

The former does look shorter, is it easier to read though?

**Comment by AlexanderSher** [comment]
It would be easier to read if we upgrade the codegen so that sync/async split is done in the REST client (since REST clients are internal types):
```cs
private async IAsyncEnumerable<Page<string>> GetRelationshipsImplAsync(bool async, string digitalTwinId, string relationshipName, [EnumeratorCancellation] CancellationToken cancellationToken)
{
    Argument.AssertNotNull(digitalTwinId, nameof(digitalTwinId));

    Response<RelationshipCollection> response = await _dtRestClient.ListRelationshipsAsync(async, digitalTwinId, relationshipName, cancellationToken).ConfigureAwait(false);
    yield return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());

    while (!string.IsNullOrEmpty(response.Value.NextLink))
    {
        response = await _dtRestClient.ListRelationshipsNextPageAsync(async, response.Value.NextLink, digitalTwinId, relationshipName, cancellationToken).ConfigureAwait(false);
        yield return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());
    }
}
```

**Comment by pakrym** [comment]
How do you support starting enumeration from the continuation token in this model?

**Comment by AlexanderSher** [comment]
> How do you support starting enumeration from the continuation token in this model?

Do you mean the case when `continuationToken` is provided to `AsPages`?

**Comment by pakrym** [comment]
Yes

**Comment by AlexanderSher** [comment]
That would require a factory method instead of instance so that when `AsPages` we can create new instance of enumerable:
```cs
private async IAsyncEnumerable<Page<string>> GetRelationshipsImplAsync(bool async, string continuationToken, string digitalTwinId, string relationshipName, [EnumeratorCancellation] CancellationToken cancellationToken)
{
    Argument.AssertNotNull(digitalTwinId, nameof(digitalTwinId));
    Response<RelationshipCollection> response;
    
    if (string.IsNullOrEmpty(continuationToken)) 
    {
        response = await _dtRestClient.ListRelationshipsAsync(async, digitalTwinId, relationshipName, cancellationToken).ConfigureAwait(false);
        continuationToken = response.Value.NextLink;
        yield return Page.FromValues(response.Value.Value, continuationToken, response.GetRawResponse());
    }

    while (!string.IsNullOrEmpty(continuationToken))
    {
        response = await _dtRestClient.ListRelationshipsNextPageAsync(async, response.Value.NextLink, digitalTwinId, relationshipName, cancellationToken).ConfigureAwait(false);
        continuationToken = response.Value.NextLink;
        yield return Page.FromValues(response.Value.Value, continuationToken, response.GetRawResponse());
    }
}
```
```cs
public virtual AsyncPageable<string> GetRelationshipsAsync(string digitalTwinId, string relationshipName = null, CancellationToken cancellationToken = default)
    => PageableHelpers.CreateAsyncPageable(continuationToken => GetRelationshipsImplAsync(true, continuationToken, digitalTwinId, relationshipName, cancellationToken),
        _clientDiagnostics, $"{nameof(DigitalTwinsClient)}.{nameof(GetRelationships)}");
```

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs [hunk 6]
```diff
@@ -253,19 +243,26 @@ Page<CustomFormModelInfo> NextPageFunc(string nextLink, int? pageSizeHint)
         /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
         /// <returns>A collection of <see cref="CustomFormModelInfo"/> items.</returns>
         [ForwardsClientCalls]
-        public virtual AsyncPageable<CustomFormModelInfo> GetCustomModelsAsync(CancellationToken cancellationToken = default)
+        public virtual Pageable<CustomFormModelInfo> GetCustomModels(CancellationToken cancellationToken = default)
+            => PageableHelpers.CreateEnumerable(GetCustomModelsImplAsync(false, cancellationToken).EnsureSyncEnumerable());
+
+        private async IAsyncEnumerable<Page<CustomFormModelInfo>> GetCustomModelsImplAsync(bool async, [EnumeratorCancellation] CancellationToken cancellationToken)
```
Comment (maririos) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs:249]: What do you mean with `Impl`?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs [hunk 7]
```diff
@@ -253,19 +243,26 @@ Page<CustomFormModelInfo> NextPageFunc(string nextLink, int? pageSizeHint)
...
         {
-            async Task<Page<CustomFormModelInfo>> FirstPageFunc(int? pageSizeHint)
-            {
-                Response<Models_internal> response = await ServiceClient.ListCustomModelsAsync(cancellationToken).ConfigureAwait(false);
-                return Page.FromValues(response.Value.ModelList.Select(info => new CustomFormModelInfo(info)), response.Value.NextLink, response.GetRawResponse());
-            }
-            async Task<Page<CustomFormModelInfo>> NextPageFunc(string nextLink, int? pageSizeHint)
+            Response<Models_internal> response = async
```
Comment (maririos) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs:251]: I personally don't mind having everything merged into one method. We've done that for other FR methods too.
In this case, not sure behavior could change more depending on `async` ...

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs (hunk 6 above)
Comment (AlexanderSher) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs:249]: Implementation

**Comment by pakrym** [comment]
Yeah, I'm reluctant to make customers to write the if and while correctly and the size saving is negligible (if any at all). But I do like the diagnostic scope simplification a lot.

**Comment by AlexanderSher** [comment]
I would argue that `if` condition / `while` loop is easier to understand than a pair of callbacks (here one can see the whole structure of the enumeration while in case of callbacks it is a kind of magic) and it is safer cause user can't introduce an undesired dependency between the closures. Understanding of concept of `continuationToken` is required in both cases.

But the main purpose of the proposed change isn't the refactoring, but to fix the issue with `WithCancellation` support. With that fix and callbacks, user code will look like this:
```cs
public virtual AsyncPageable<ModelData> GetModelsAsync(IEnumerable<string> dependenciesFor = default, bool includeModelDefinition = false, GetModelsOptions options = default, CancellationToken cancellationToken = default)
{
    async Task<Page<ModelData>> FirstPageFunc(int? pageSizeHint, CancellationToken ct)
    {
        Response<PagedModelDataCollection> response = await _dtModelsRestClient.ListAsync(dependenciesFor, includeModelDefinition, options, ct).ConfigureAwait(false);
        return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());
    }

    async Task<Page<ModelData>> NextPageFunc(string nextLink, int? pageSizeHint, CancellationToken ct)
    {
        Response<PagedModelDataCollection> response = await _dtModelsRestClient.ListNextPageAsync(nextLink, dependenciesFor, includeModelDefinition, options, ct).ConfigureAwait(false);
        return Page.FromValues(response.Value.Value, response.Value.NextLink, response.GetRawResponse());
    }

    return PageableHelpers.CreateAsyncEnumerable(FirstPageFunc, NextPageFunc, cancellationToken);
}
```

With two `CancellationToken` in scope, it is very easy to accidentally use the wrong one in the REST client method calls

**Comment by jsquire** [comment]
@AlexanderSher:  Would you please advise as to whether this is still an active workstream or whether this can be closed out?

**Comment by AlexanderSher** [comment]
This change can be applied only after related changes in codegen. @pakrym , do you know if we have a task for it?

**Comment by jsquire** [comment]
Thanks, @AlexanderSher.   I'll untag and let it live on.

**Comment by pakrym** [comment]
I could've sworn there was one but didn't find any. Filed https://github.com/Azure/autorest.csharp/issues/906

**Comment by jsquire** [comment]
I'm going to close this out, since the associated AutoRest issue is still on the backlog 6 months later.   Please feel free to reopen when you're unblocked and ready to restart work on this.

**Pull request closed without merging**